k2/python/tests/generate_denominator_lattice_test.py

```diff
@@ -0,0 +1,216 @@
+
+    index = (
+        torch.arange(S, device=src.device)
+        .view((1, S))
+        .repeat((T, 1))
+        .repeat((B, 1, 1))
+    )
+    index = (index - shifts.reshape(B, T, 1)) % S
+    return torch.gather(src, 2, index)
+
+
+def simulate_importance_sampling(
+    batch_size: int,
+    vocab_size: int,
+    path_length: int,
+    num_paths: int,
+    context_size: int = 2,
+    blank_id: int = 0,
+    device: torch.device = torch.device("cpu"),
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Args:
+      batch_size:
+        The number of sequence.
+      vocab_size:
+        Vocabulary size.
+      path_length:
+        How many symbols we will sample for each path.
+      num_paths:
+        How many paths we will sample for each sequence.
+      context_size:
+        The number of left symbols.
+      blank_id:
+        Stands for null context.
+
+    Returns:
+      Three tensors will be returned.
+      - sampled_paths:
+        A tensor of shape (batch_size, num_paths, path_length), containing the
+        sampled symbol ids.
+      - sampling_probs:
+        A tensor of shape (batch_size, num_paths, path_length), containing the
+        sampling probabilities of the sampled symbols.
+      - left_symbols:
+        A tensor of shape (batch_size, num_paths, path_length, context_size),
+        containing the left symbols of the sampled symbols.
+      - frame_ids:
+        A tensor of shape (batch_size, num_paths, path_length), containing the
+        frame ids at which we sampled the symbols.
+    """
+    # we sample paths from frame 0
+    t_index = torch.zeros(
+        (batch_size, num_paths), dtype=torch.int64, device=device
+    )
+
+    left_symbols = torch.tensor(
+        [blank_id], dtype=torch.int64, device=device
+    ).expand(batch_size, num_paths, context_size)
+
+    sampled_paths_list = []
+    sampling_probs_list = []
+    frame_ids_list = []
+    left_symbols_list = []
+
+    for i in range(path_length):
+        probs = torch.randn(batch_size, num_paths, vocab_size)
+        probs = torch.softmax(probs, -1)
+        # sampler: https://pytorch.org/docs/stable/distributions.html#categorical
+        sampler = Categorical(probs=probs)
+
+        # sample one symbol for each path
+        # index : (batch_size, num_paths)
+        index = sampler.sample()
+        sampled_paths_list.append(index)
+
+        # gather sampling probabilities for corresponding indexs
+        # sampling_prob : (batch_size, num_paths, 1)
+        sampling_probs = torch.gather(probs, dim=2, index=index.unsqueeze(2))
+        sampling_probs_list.append(sampling_probs.squeeze(2))
+
+        frame_ids_list.append(t_index)
+
+        left_symbols_list.append(left_symbols)
+
+        # update (t, s) for each path
+        # index == 0 means the sampled symbol is blank
+        t_mask = index == 0
+        #         t_index = torch.where(t_mask, t_index + 1, t_index)
+        t_index = t_index + 1
```

Comment:
If we use regular RNN-T, it is possible to generate too many symbols for a specific frame, and that might be chances to generate a lattice containing cycles, which is not expected.  I am not sure whether we will encounter such a issue at the very beginning of training.

Reply:
> 

Hm, a valid point.  Yes, computing forward backward scores would not work correctly if there are cycles.  One possibility would be to augment the state with a sub-frame, i.e. instead of (ctx, t) it becomes (ctx, t, sub_t) with sub_t = (0, 1, 2, ..).  That would prevent cycles, although it might prevent a small number of paths from recombining that might otherwise be able to recombine.